[TEP-0112] Replace volumes with Workspaces [Proposal]

lbernick opened the pull request:
This TEP proposes removing volumes from the Task spec in the v1 API, as well as removing
volumeMounts and volumeDevices for the container-like fields (Step, StepTemplate, and Sidecar).

/kind tep

lbernick commented:
Addresses https://github.com/tektoncd/pipeline/issues/2058

@vdemeester @skaegi PTAL

lbernick commented:
/hold

need to do a bit more digging--will revisit the proposal to support hostPath
@wlynch relevant for you as well

lbernick commented:
/hold cancel

lbernick commented:
/assign @dibyom @pierretasci

pierretasci commented on teps/0111-remove-volumes-from-task-spec.md:
```diff
@@ -0,0 +1,252 @@
+1. Update examples our documentation to demonstrate the use of workspaces instead of volumes for the use cases listed in
+this TEP.
+  - The v1beta1 -> v1 migration guide should include examples of migrating volumes to workspaces.
+1. Deprecate `Task.Spec.Volumes`, `Step.VolumeMounts`, and `Step.VolumeDevices` (and likewise for `StepTemplate` and `Sidecar`).
```
We have another use case not already covered here and not supported by workspaces (nor could they easily be) which is dynamic volumes. Sometimes, it is desired/opportune to create a volume (say backed by a configmap or secret) dynamically in a pipeline and then mount it into an already named volume mount in a task. This is not possible with workspaces because the workspace bindings need to be defined up front.

lbernick replied:
Thanks Pierre! I added this as a use case that workspaces don't currently support, and modified the TEP to specify that before it's merged as implementable, we should have a plan for workspaces support for this use case. PTAL!

lbernick commented:
I chatted with @pierretasci yesterday about the "dynamic volumes" use case, and added it here as an existing use case that workspaces don't support. I need to think more about what we should do about this, but for now I added "dynamic workspaces" as an optional workspace enhancement. We are both in agreement that this is the right long-term direction for volumes/workspaces, but want to avoid removing support for an existing use case even if it isn't an "intended" use of the API.

I've also added more detail on the options for different volume types we might want to support in workspaces, including embedding the k8s volumesource in workspacebinding. For now I think the existing workspace binding options are sufficient for the use cases we're targeting.

afrittoli commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,350 @@
+by Workspaces. However, this approach is not recommended for
+[security reasons](https://blog.quarkslab.com/kubernetes-and-hostpath-a-love-hate-relationship.html).
+
+The only Catalog Task using hostPath volumes is [`kind`](https://hub.tekton.dev/tekton/task/kind).
+For more information on why `kind` on Kubernetes requires hostPath volumes, see
+https://github.com/kubernetes-sigs/kind/issues/303#issuecomment-521384993 and
+[Running KIND Inside A Kubernetes Cluster For Continuous Integration](https://d2iq.com/blog/running-kind-inside-a-kubernetes-cluster-for-continuous-integration)
+(TL;DR: because `kind` needs to create nested Docker containers).
```
What would be the alternative here if we removed workspaces?
I agree that using volumes directly instead of workspaces is not a best practice, but unless we have an alternative here I wonder if we could address this by documentation, i.e. make sure that volumes are not used in any example / catalog task unless were strictly required. We could add a special page / section to document this feature as something we do not recommend to use unless in specific documented cases and specify security risks.

lbernick replied:
I've added this as an option to "alternatives". Right now, I don't think we would have an alternative for this using workspaces.

lbernick commented:
@pritidesai @afrittoli @pierretasci thanks for the feedback so far!
I think there are a few too many open questions so I've marked this as "proposed", removed the "proposal" section, and put the ideas being floated under "alternatives" (can always add more). Let's align on the use cases we're addressing and then we can figure out what to do about it.

lbernick commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,350 @@
+  workspaces:
+  - name: docker-socket
+    mountPath: /var/run/docker.sock
+    emptyDir: {}
```
nice idea, I added this to alternatives as well and put pros + cons for both solutions.

dibyom commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,368 @@
+Pros:
+- Supports common use cases of caching and sharing data with sidecars, without requiring `Task` users to understand `Task` implementation
+
+Cons:
```
Ah, looks like this is the next alternative :)

dibyom commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,368 @@
+
+Cons:
+- Not necessary for most CI/CD use cases.
+- Tasks that require hostPath would likely want to declare this dependency at authoring time rather than at runtime.
```
Looking at both the internal and this hostPath use case, it seems like there are aa few use cases where the task authors need to provide the volume/workspace type they need and workspaces today only allow taskrun authors to supply these

lbernick replied:
agreed! I think this is covered under the "task-internal" workspaces option

Tomcli commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,368 @@
+### Use TaskRun.Spec.PodTemplate to support missing Workspace features provided by Volumes
+
+A [proposed alternative](https://github.com/tektoncd/pipeline/issues/2057#issuecomment-587559783)
+to implementing support for extra volume types as workspace bindings is to use the `podTemplate` field
```
If we are going to remove volumemount and volumes for tasks, will there be a document on how to migrate podTemplate.volumes with workspaces?

lbernick replied:
Definitely! I added a section about this under "design considerations".

lbernick commented on teps/0111-remove-volumes-from-task-spec.md:
```diff
@@ -0,0 +1,252 @@
+
+These secrets can also be isolated to only the Steps that need them using
+[isolated workspaces](https://tekton.dev/docs/pipelines/workspaces/#isolating-workspaces-to-specific-steps-or-sidecars)
+(note: currently in alpha).
```
resolving since this is now in "proposed" state with proposal removed

lbernick commented on teps/0111-replace-volumes-with-workspaces.md:
```diff
@@ -0,0 +1,368 @@
+
+This TEP proposes removing Volumes from the Task spec, and removing VolumeMounts and VolumeDevices from Step,
+StepTemplate, and Sidecar. Users are encouraged to use Workspaces as a replacement feature.
+
```
done! moved under "proposal"

pierretasci commented:
/lgtm
